File: lib/src/main/scala/io/github/kory33/guardedqueries/core/utils/extensions/ConjunctiveQueryExtensions.java
package io.github.kory33.guardedqueries.core.utils.extensions;

import com.google.common.collect.ImmutableSet;
import io.github.kory33.guardedqueries.core.utils.algorithms.SimpleUnionFindTree;
import uk.ac.ox.cs.pdq.fol.Atom;
import uk.ac.ox.cs.pdq.fol.ConjunctiveQuery;
import uk.ac.ox.cs.pdq.fol.Constant;
import uk.ac.ox.cs.pdq.fol.Variable;

import java.util.Arrays;
import java.util.Collection;
import java.util.Optional;
import java.util.function.Predicate;
import java.util.stream.Stream;

public class ConjunctiveQueryExtensions {
    private ConjunctiveQueryExtensions() {
    }

    /**
     * Computes a subquery of a given {@code ConjunctiveQuery} that includes only the atoms
     * satisfying a specified predicate.
     * <p>
     * Since {@code ConjunctiveQuery} cannot be empty, an empty {@code Optional} is returned
     * if the predicate is not satisfied by any atom in the given {@code ConjunctiveQuery}.
     */
    public static Optional<ConjunctiveQuery> filterAtoms(
            final ConjunctiveQuery conjunctiveQuery,
            final Predicate<? super Atom> atomPredicate
    ) {
        final var originalFreeVariables = ImmutableSet.copyOf(conjunctiveQuery.getFreeVariables());

        final var filteredAtoms = Arrays.stream(conjunctiveQuery.getAtoms())
                .filter(atomPredicate)
                .toArray(Atom[]::new);

        // variables in filteredAtoms that are free in the original conjunctiveQuery
        final var filteredFreeVariables = Arrays.stream(filteredAtoms)
                .flatMap(atom -> Arrays.stream(atom.getVariables()))
                .filter(originalFreeVariables::contains)
                .toArray(Variable[]::new);

        if (filteredAtoms.length == 0) {
            return Optional.empty();
        } else {
            return Optional.of(ConjunctiveQuery.create(filteredFreeVariables, filteredAtoms));
        }
    }

    /**
     * Computes a subquery of a given Conjunctive Query that includes only the atoms
     * that contains at least one bound variable and all bound variables in the atom
     * are present in a specified set of variables.
     * <p>
     * The set of variables in the returned subquery is a subset of {@code variables},
     * and a variable is bound in the returned subquery if and only if it is bound in {@code conjunctiveQuery}.
     * <p>
     * For example, if {@code conjunctiveQuery} is {@code ∃x,y,z. T(x,y,z) ∧ T(x,y,w) ∧ T(x,c,z)}
     * and {@code boundVariableSet} is {@code {x,y}}, then the returned subquery is
     * {@code ∃x,y. T(x,y,w)}.
     * <p>
     * If no atom in the given {@code ConjunctiveQuery} has variable set entirely contained in
     * {@code variables}, an empty {@code Optional} is returned.
     *
     * @param conjunctiveQuery The Conjunctive Query to compute the subquery from
     * @param variables        The filter of variables that should be included in the subquery.
     * @return The computed subquery.
     */
    public static Optional<ConjunctiveQuery> strictlyInduceSubqueryByVariables(
            final ConjunctiveQuery conjunctiveQuery,
            final Collection<? extends Variable> variables
    ) {
        final var variableSet = ImmutableSet.<Variable>copyOf(variables);
        final var cqBoundVariables = ImmutableSet.copyOf(conjunctiveQuery.getBoundVariables());

        return filterAtoms(conjunctiveQuery, atom -> {
            // variables in the atom that are bound in the CQ
            final var atomBoundVariables = SetLikeExtensions.intersection(
                    Arrays.asList(atom.getVariables()),
                    cqBoundVariables
            );
            return variableSet.containsAll(atomBoundVariables) && atomBoundVariables.size() > 0;
        });
    }

    /**
     * Computes a subquery of a given Conjunctive Query that includes only the atoms
     * which have at least one bound variable in a specified set of variables.
     * <p>
     * For example, if {@code conjunctiveQuery} is {@code ∃x,y,z. T(x,y,w) ∧ T(x,c,z)}
     * and {@code boundVariableSet} is {@code {y}}, then the returned subquery is
     * {@code ∃x,y. T(x,y,w)}.
     * <p>
     * If no atom in the given {@code ConjunctiveQuery} has variable set intersecting with
     * {@code variables}, an empty {@code Optional} is returned.
     */
    public static Optional<ConjunctiveQuery> subqueryRelevantToVariables(
            final ConjunctiveQuery conjunctiveQuery,
            final Collection<? extends Variable> variables
    ) {
        final var variableSet = ImmutableSet.<Variable>copyOf(variables);
        final var cqBoundVariables = ImmutableSet.copyOf(conjunctiveQuery.getBoundVariables());

        return filterAtoms(conjunctiveQuery, atom -> {
            // variables in the atom that are bound in the CQ
            final var atomBoundVariables = SetLikeExtensions.intersection(
                    Arrays.asList(atom.getVariables()),
                    cqBoundVariables
            );
            return SetLikeExtensions.nontriviallyIntersects(atomBoundVariables, variableSet);
        });
    }

    /**
     * Variables in the strict neighbourhood of a given set of variables in the given CQ.
     * <p>
     * Given a conjunctive query {@code q} and a variable {@code x} appearing in {@code q},
     * {@code x} is said to be in the strict neighbourhood of a set {@code V} of variables if
     * <ol>
     *   <li>{@code x} is not an element of {@code V}, and</li>
     *   <li>{@code x} occurs in the subquery of {@code q} relevant to {@code V}.</li>
     * </ol>
     */
    public static ImmutableSet<Variable> neighbourhoodVariables(
            final ConjunctiveQuery conjunctiveQuery,
            final Collection<? extends Variable> variables
    ) {
        final var subquery = subqueryRelevantToVariables(conjunctiveQuery, variables);
        if (subquery.isEmpty()) {
            return ImmutableSet.of();
        }

        return SetLikeExtensions.difference(variablesIn(subquery.get()), variables);
    }

    /**
     * Given a conjunctive query {@code conjunctiveQuery} and a set {@code boundVariables} of variables
     * bound in {@code conjunctiveQuery}, returns a stream of all {@code conjunctiveQuery}-connected
     * components of {@code variables}.
     */
    public static Stream</* nonempty */ImmutableSet<Variable>> connectedComponents(
            final ConjunctiveQuery conjunctiveQuery,
            final Collection<? extends Variable> boundVariables
    ) {
        if (boundVariables.isEmpty()) {
            return Stream.empty();
        }

        for (final var variable : boundVariables) {
            if (!Arrays.asList(conjunctiveQuery.getBoundVariables()).contains(variable)) {
                throw new IllegalArgumentException("Variable " + variable + " is not bound in the given CQ");
            }
        }

        final var unionFindTree = new SimpleUnionFindTree<Variable>(boundVariables);
        for (final var atom : conjunctiveQuery.getAtoms()) {
            final var variablesToUnion = SetLikeExtensions.intersection(
                    ImmutableSet.copyOf(atom.getVariables()),
                    boundVariables
            );
            unionFindTree.unionAll(variablesToUnion);
        }
        return unionFindTree.getEquivalenceClasses().stream();
    }

    /**
     * Given a conjunctive query {@code q} and a set {@code v} of variables in {@code q},
     * checks if {@code v} is connected in {@code q}.
     * <p>
     * A set of variables {@code V} is said to be connected in {@code q} if
     * there is at most one {@code q}-connected component of {@code V} in {@code q}.
     */
    public static boolean isConnected(
            final ConjunctiveQuery conjunctiveQuery,
            final Collection<? extends Variable> variables
    ) {
        return connectedComponents(conjunctiveQuery, variables).count() <= 1L;
    }

    public static ImmutableSet<Constant> constantsIn(
            final ConjunctiveQuery conjunctiveQuery
    ) {
        return ImmutableSet.copyOf(
                StreamExtensions
                        .filterSubtype(Arrays.stream(conjunctiveQuery.getTerms()), Constant.class)
                        .iterator()
        );
    }

    public static ImmutableSet<Variable> variablesIn(
            final ConjunctiveQuery conjunctiveQuery
    ) {
        return SetLikeExtensions.union(
                Arrays.asList(conjunctiveQuery.getBoundVariables()),
                Arrays.asList(conjunctiveQuery.getFreeVariables())
        );
    }
}

File: lib/src/main/scala/io/github/kory33/guardedqueries/core/utils/extensions/ConjunctiveQueryExtensions.scala
package io.github.kory33.guardedqueries.core.utils.extensions

import io.github.kory33.guardedqueries.core.utils.extensions.IterableExtensions.given
import io.github.kory33.guardedqueries.core.utils.algorithms.SimpleUnionFindTree
import uk.ac.ox.cs.pdq.fol.Atom
import uk.ac.ox.cs.pdq.fol.ConjunctiveQuery
import uk.ac.ox.cs.pdq.fol.Constant
import uk.ac.ox.cs.pdq.fol.Variable

object ConjunctiveQueryExtensions {
  // We enclose the extension methods in the given instance; see
  // https://github.com/lampepfl/dotty/issues/14777#issuecomment-1078995983 for details
  given Extensions: AnyRef with
    extension (conjunctiveQuery: ConjunctiveQuery)
      /**
       * Computes a subquery of a given `ConjunctiveQuery` that includes only the atoms
       * satisfying a specified predicate.
       *
       * Since `ConjunctiveQuery` cannot be empty, an empty `Optional` is returned if the
       * predicate is not satisfied by any atom in the given `ConjunctiveQuery`.
       */
      def filterAtoms(atomPredicate: Atom => Boolean): Option[ConjunctiveQuery] = {
        val originalFreeVariables = conjunctiveQuery.getFreeVariables.toSet
        val filteredAtoms = conjunctiveQuery.getAtoms.filter(atomPredicate)

        if (filteredAtoms.nonEmpty) {
          // variables in filteredAtoms that are free in the original conjunctiveQuery
          val filteredFreeVariables = filteredAtoms
            .flatMap(_.getVariables)
            .filter(originalFreeVariables.contains)

          Some(ConjunctiveQuery.create(filteredFreeVariables, filteredAtoms))
        } else {
          None
        }
      }

      /**
       * Computes a subquery of a given Conjunctive Query that includes only the atoms that
       * contains at least one bound variable and all bound variables in the atom are present in
       * a specified set of variables.
       *
       * The set of variables in the returned subquery is a subset of `variables`, and a
       * variable is bound in the returned subquery if and only if it is bound in
       * `conjunctiveQuery`.
       *
       * For example, if `conjunctiveQuery` is `∃x,y,z. T(x,y,z) ∧ T(x,y,w) ∧ T(x,c,z)` and
       * `boundVariableSet` is `{x,y`}, then the returned subquery is `∃x,y. T(x,y,w)`.
       *
       * If no atom in the given `ConjunctiveQuery` has variable set entirely contained in
       * `variables`, an empty `Optional` is returned.
       *
       * @param variables
       *   The filter of variables that should be included in the subquery.
       * @return
       *   The computed subquery.
       */
      def strictlyInduceSubqueryByVariables(variables: Set[Variable])
        : Option[ConjunctiveQuery] = {
        val queryBoundVariables = conjunctiveQuery.getBoundVariables.toSet
        conjunctiveQuery.filterAtoms(atom => {
          // variables in the atom that are bound in the CQ
          val atomBoundVariables = atom.getVariables.toSet.intersect(queryBoundVariables)
          atomBoundVariables.subsetOf(variables) && atomBoundVariables.nonEmpty
        })
      }

      /**
       * Computes a subquery of a given Conjunctive Query that includes only the atoms which
       * have at least one bound variable in a specified set of variables.
       *
       * For example, if `conjunctiveQuery` is `∃x,y,z. T(x,y,w) ∧ T(x,c,z)` and
       * `boundVariableSet` is `{y`}, then the returned subquery is `∃x,y. T(x,y,w)`.
       *
       * If no atom in the given `ConjunctiveQuery` has variable set intersecting with
       * `variables`, an empty `Optional` is returned.
       */
      def subqueryRelevantToVariables(variables: Set[Variable]): Option[ConjunctiveQuery] = {
        val queryBoundVariables = conjunctiveQuery.getBoundVariables.toSet
        conjunctiveQuery.filterAtoms(atom => {
          // variables in the atom that are bound in the CQ
          val atomBoundVariables = atom.getVariables.toSet.intersect(queryBoundVariables)
          atomBoundVariables.intersects(variables)
        })
      }

      /**
       * Variables in the strict neighbourhood of a given set of variables in the given CQ.
       *
       * Given a conjunctive query `q` and a variable `x` appearing in `q`, `x` is said to be in
       * the strict neighbourhood of a set `V` of variables if <ol> <li>`x` is not an element of
       * `V`, and</li> <li>`x` occurs in the subquery of `q` relevant to `V`.</li> </ol>
       */
      def strictNeighbourhoodOf(variables: Set[Variable]): Set[Variable] =
        conjunctiveQuery.subqueryRelevantToVariables(variables)
          .map { subquery => subquery.allVariables -- variables }
          .getOrElse(Set.empty)

      /**
       * Given a conjunctive query `conjunctiveQuery` and a set `boundVariables` of variables
       * bound in `conjunctiveQuery`, returns a stream of all `conjunctiveQuery`-connected
       * components of `variables`.
       */
      def connectedComponentsOf(boundVariables: Set[Variable]): Set[Set[Variable]] = {
        if (boundVariables.isEmpty) return Set()

        for (variable <- boundVariables) {
          if (!conjunctiveQuery.getBoundVariables.contains(variable))
            throw new IllegalArgumentException(
              s"Variable $variable is not bound in the given CQ $conjunctiveQuery"
            )
        }

        val unionFindTree = SimpleUnionFindTree(boundVariables)
        for (atom <- conjunctiveQuery.getAtoms) {
          unionFindTree.unionAll(atom.getVariables.toSet.intersect(boundVariables))
        }
        unionFindTree.getEquivalenceClasses
      }

      /**
       * Given a conjunctive query `q` and a set `v` of variables in `q`, checks if `v` is
       * connected in `q`.
       *
       * A set of variables `V` is said to be connected in `q` if there is at most one
       * `q`-connected component of `V` in `q`.
       */
      def connects(variables: Set[Variable]): Boolean =
        conjunctiveQuery.connectedComponentsOf(variables).size <= 1

      def allConstants: Set[Constant] =
        conjunctiveQuery.getTerms.collect { case constant: Constant => constant }.toSet

      def allVariables: Set[Variable] =
        conjunctiveQuery.getBoundVariables.toSet.union(conjunctiveQuery.getFreeVariables.toSet)
}
